### PR TITLE
[CALCITE-3408] Add support for enumerable intersect/minus all

### PR DESCRIPTION
**Pull request by yanlin-Lynn**
Currently, [EnumerableIntersect](https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableIntersect.java#L37) and [EnumerableMinus](https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableMinus.java#L37) with true value for `all` is not supported.
Thus, exception occurs when trying to execute sql contains `intersect all`  or `except all`, please refer to [CALCITE-3408](https://issues.apache.org/jira/browse/CALCITE-3408) for more details.

This PR tries to add support for enumerable intersect/minus all.

### CODE REVIEW

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java [hunk 1]
```diff
@@ -532,15 +533,19 @@ public void remove() {
    * by Enumerable.)
    */
   public static <TSource> Enumerable<TSource> except(
-      Enumerable<TSource> source0, Enumerable<TSource> source1) {
+      Enumerable<TSource> source0, Enumerable<TSource> source1, boolean all) {
     Set<TSource> set = new HashSet<>();
-    source0.into(set);
-    try (Enumerator<TSource> os = source1.enumerator()) {
+    Collection<TSource> resultCollection = all ? HashMultiset.create() : new HashSet<>();
```
Comment (amaliujia) [inline linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Is there a source of definition on `except/intersect all` implementation?

**Comment by amaliujia** [comment]
cc @amaliujia

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 1 above)
Comment (yanlin-Lynn) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: hi, @amaliujia I'm not quite sure what do you want to ask, and I guess you were asking the implementation of `EnumerableIntersect` and `EnumerableMinus`.
They are are in [EnumerableIntersect](https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableIntersect.java#L45) and [EnumerableMinus](https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableMinus.java#L45)

I update the logic of these to support `all`.

Comment (amaliujia) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Thank Yanlin. 

I was looking for some preciously definition of the set op "ALL" implementation (like comments). 

Basically either something like:
```
              // Say for Row R, there are m instances on left and n instances on right,
              // INTERSECT ALL outputs MIN(m, n) instances of R.
```

Or 
```
Calcite implementation follows XX DB's implementation. See their doc(link)
```

If there is no such thing I will just read code.

Comment (yanlin-Lynn) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: I haven't found that.

**Comment by julianhyde** [comment]
I think you should keep the old methods, not deprecated. Treat 'all' as an optional parameter whose default is false.

Overloaded methods with different parameters should have different javadoc descriptions. It was an oversight not to change the description when you added the 'all' parameter. Can you fix it.

**Comment by yanlin-Lynn** [comment]
Sure, I'll update the PR.

**Comment by yanlin-Lynn** [comment]
Thanks for review, I'll delete the branch.

**Pull request closed without merging**